lib/dns-resolver.js

```diff
@@ -0,0 +1,516 @@
+      if (iface.internal) {
+        continue
+      }
+
+      if (iface.family === 'IPv6') {
+        has6 = true
+      } else {
+        has4 = true
+      }
+
+      if (has4 && has6) {
+        return { has4, has6 }
+      }
+    }
+  }
+
+  return { has4, has6 }
+}
+
+const isIterable = (map) => {
+  return Symbol.iterator in map
+}
+
+const ignoreNoResultErrors = (dnsPromise) => {
+  return dnsPromise.catch((error) => {
+    if (
+      error.code === 'ENODATA' ||
+      error.code === 'ENOTFOUND' ||
+      error.code === 'ENOENT' // Windows: name exists, but not this record type
+    ) {
+      return []
+    }
+
+    throw error
+  })
+}
+
+const ttl = { ttl: true }
+const all = { all: true }
+const all4 = { all: true, family: 4 }
+const all6 = { all: true, family: 6 }
+
+class DNSResolver {
+  #resolver
+  #cache
+  #dnsLookup
+  #iface
+  #pending = {}
+  #nextRemovalTime = false
+  #removalTimeout
+  #hostnamesToFallback = new Set()
+  #fallbackInterval
+  #resolve4
+  #resolve6
+
+  constructor ({
+    cache = new Map(),
+    maxTtl = Infinity,
+    fallbackDuration = 3600,
+    errorTtl = 0.15,
+    resolver = new AsyncResolver(),
+    lookup = dnsLookup,
+    lookupOptions = undefined,
+    roundRobinStrategy = 'first'
+  } = {}) {
+    this.maxTtl = maxTtl
+    this.errorTtl = errorTtl
+
+    if (roundRobinStrategies.includes(roundRobinStrategy) === false) {
+      throw new Error(`roundRobinStrategy must be one of: ${roundRobinStrategies.join(', ')}`)
+    }
+
+    this.roundRobinStrategy = roundRobinStrategy
+
+    this.#cache = cache
+    this.#resolver = resolver
+    this.#dnsLookup = lookup && promisify(lookup)
+    this.stats = {
+      cache: 0,
+      query: 0
+    }
+
+    if (this.#resolver instanceof AsyncResolver) {
+      this.#resolve4 = this.#resolver.resolve4.bind(this.#resolver)
+      this.#resolve6 = this.#resolver.resolve6.bind(this.#resolver)
+    } else {
+      this.#resolve4 = promisify(this.#resolver.resolve4.bind(this.#resolver))
+      this.#resolve6 = promisify(this.#resolver.resolve6.bind(this.#resolver))
+    }
+
+    this.#iface = getIfaceInfo()
+
+    this.#pending = {}
+    this.#nextRemovalTime = false
+    this.#hostnamesToFallback = new Set()
+
+    this.fallbackDuration = fallbackDuration
+
+    if (fallbackDuration > 0) {
+      const interval = setInterval(() => {
+        this.#hostnamesToFallback.clear()
+      }, fallbackDuration * 1000)
+
+      /* istanbul ignore next: There is no `interval.unref()` when running inside an Electron renderer */
+      if (interval.unref) {
+        interval.unref()
+      }
+
+      this.#fallbackInterval = interval
+    }
+
+    if (lookupOptions) {
+      this.lookup = (hostname, _options, callback) => this.#lookup(hostname, lookupOptions, callback)
+    } else {
+      this.lookup = this.#lookup.bind(this)
+    }
+    this.lookupAsync = this.lookupAsync.bind(this)
+  }
+
+  get _cache () {
+    return this.#cache
+  }
```

Comment:
from the tests, it looks like this is only needed to check how many entries are in the cache. Would it make sense to only expose the number of cached entries, rather than the cache itself? If not, changing these to a symbol property to deter users from using this would be better.

Reply:
Yes agree

Reply:
Done